pkg/codegen/go/gen.go

```diff
@@ -401,7 +407,15 @@ func (pkg *pkgContext) typeStringImpl(t schema.Type, argsType bool) string {
 		typ := "map[string]"
 		return typ + pkg.typeStringImpl(t.ElementType, argsType)
 	case *schema.ObjectType:
-		return pkg.resolveObjectType(t)
+		typ := pkg.resolveObjectType(t)
+		// handle top-level/namespaceless args
+		if argsType && !strings.Contains(typ, ".") {
+			prefix := pkg.getImportPrefix(t.Token, t)
+			if prefix != "" {
+				return prefix + "." + typ
+			}
+		}
```

Comment:
Feels like we could be more precise here: we'll need the prefix if the referring module does not match the referent module, so it seems like that should be the condition (rather than `argsType && !strings.Contains(top, ".")`).

Reply:
So, I agree with you here but my perception of the problem is that we have intertwined the SDK generation and program gen codepaths in a very complicated and brittle way. 

I attempted to make the fix at this layer originally by reversing these two if statements:

https://github.com/pulumi/pulumi/blob/97183b83a38385f06d3bbd4ff6999eeec2e5624e/pkg/codegen/go/gen.go#L1835-L1840

However this codepath is shared with the SDK generator and breaks all SDK gen for `pulumiTypes`. The declared types become `type aws.Foo struct`...

At this layer there is also no context from the calling package (`main`).